File: rust/sdk/src/builders/action.rs
use crate::account::AccountId;
use crate::builders::{TxnFilter, DEFAULT_TXN_LIMIT};
use crate::types::Target;
use crate::WithContext;
use core::convert::From;
use m10_protos::prost::Message;
use m10_protos::sdk::list_actions_request::Filter;
use m10_protos::{sdk, Metadata};

pub struct ActionBuilder {
    name: String,
    from: AccountId,
    to: Target,
    payload: Vec<u8>,
}

impl ActionBuilder {
    pub fn for_account(name: impl Into<String>, from: AccountId, to: AccountId) -> Self {
        Self {
            name: name.into(),
            from,
            to: Target::Account(to),
            payload: vec![],
        }
    }

    pub fn for_all(name: impl Into<String>, from: AccountId) -> Self {
        Self {
            name: name.into(),
            from,
            to: Target::Any,
            payload: vec![],
        }
    }

    pub fn metadata(mut self, value: impl Metadata) -> Self {
        self.payload = value.any().encode_to_vec();
        self
    }

    pub fn payload(mut self, payload: Vec<u8>) -> Self {
        self.payload = payload;
        self
    }
}

impl From<ActionBuilder> for sdk::InvokeAction {
    fn from(action: ActionBuilder) -> Self {
        Self {
            from_account: action.from.to_vec(),
            name: action.name,
            target: Some(action.to.into()),
            payload: action.payload,
        }
    }
}

#[derive(Clone, Debug)]
pub struct ActionsFilter {
    name: String,
    filter: Filter,
}

impl TxnFilter<ActionsFilter> {
    pub fn by_account(name: String, id: AccountId) -> Self {
        Self {
            filter: ActionsFilter {
                name,
                filter: Filter::AccountId(id.to_vec()),
            },
            min: 0,
            max: u64::MAX,
            limit: DEFAULT_TXN_LIMIT,
        }
    }

    pub fn by_context_id(name: String, context_id: Vec<u8>) -> Self {
        Self {
            filter: ActionsFilter {
                name,
                filter: Filter::ContextId(context_id),
            },
            min: 0,
            max: u64::MAX,
            limit: DEFAULT_TXN_LIMIT,
        }
    }
}

impl From<TxnFilter<ActionsFilter>> for sdk::ListActionsRequest {
    fn from(filter: TxnFilter<ActionsFilter>) -> Self {
        sdk::ListActionsRequest {
            name: filter.filter.name,
            filter: Some(filter.filter.filter),
            min_tx_id: filter.min,
            max_tx_id: filter.max,
            limit: filter.limit,
        }
    }
}

impl WithContext for ActionBuilder {}

File: node/m10-sdk/src/builders/action.ts
import { m10 } from "../../protobufs";
import type { TxId } from "../ids";
import type { AccountId } from "../ids";
import { Target } from "../types";

import type { ITxnFilter } from "./txn_filter";
import { TxnFilter } from "./txn_filter";


export class ActionBuilder {

    private _name: string;
    private _from: AccountId;
    private _to: Target;
    private _payload?: Option<Uint8Array>;


    private constructor(name: string, from: AccountId, to: Target, payload?: Uint8Array) {
        this._name = name;
        this._from = from;
        this._to = to;
        this._payload = payload;
    }

    public static forAccount(name: string, from: AccountId, to: AccountId): ActionBuilder {
        return new ActionBuilder(
            name,
            from,
            new Target(to),
        );
    }

    public static forAll(name: string, from: AccountId): ActionBuilder {
        return new ActionBuilder(
            name,
            from,
            new Target(),
        );
    }

    // TODO: find usecase
    // public metadata(value: google.protobuf.Any): ActionBuilder {
    //     this._payload = ...
    //     return this;
    // }

    public payload(payload: Uint8Array): ActionBuilder {
        this._payload = payload;
        return this;
    }


    public toInvokeAction(): m10.sdk.transaction.InvokeAction {
        return new m10.sdk.transaction.InvokeAction({
            name: this._name,
            fromAccount: this._from.toUint8Array(),
            target: this._to.into(),
            payload: this._payload,
        });
    }
}

export interface IActionsFilter extends ITxnFilter {
    min: TxId;
    max: TxId;
    limit: number;
    name: string;
    accountId: Option<AccountId>;
    contextId: Option<Uint8Array>;
}

export class ActionsFilter extends TxnFilter {

    private _name: string;
    private _accountId: Option<AccountId>;
    private _contextId: Option<Uint8Array>;


    public constructor(properties: IActionsFilter) {
        super(properties);

        this._name = properties.name;
        this._accountId = properties.accountId;
        this._contextId = properties.contextId;
    }


    public min(txId: TxId): ActionsFilter {
        super.min(txId);
        return this;
    }

    public max(txId: TxId): ActionsFilter {
        super.max(txId);
        return this;
    }

    public limit(count: number): ActionsFilter {
        super.limit(count);
        return this;
    }

    public toListActionsRequest(): m10.sdk.transaction.ListActionsRequest {
        return new m10.sdk.transaction.ListActionsRequest({
            name: this._name,
            minTxId: this.getMinTxId(),
            maxTxId: this.getMaxTxId(),
            limit: this.getLimit(),
            accountId: this._accountId?.toUint8Array(),
            contextId: this._contextId,
        });
    }

    public static byAccount(name: string, accountId: AccountId): ActionsFilter {
        return new ActionsFilter({
            min: 0,
            max: Number.MAX_SAFE_INTEGER,
            limit: TxnFilter.DEFAULT_TXN_LIMIT,
            name: name,
            accountId: accountId,
            contextId: null,
        });
    }

    public static byContextId(name: string, contextId: Uint8Array): ActionsFilter {
        return new ActionsFilter({
            min: 0,
            max: Number.MAX_SAFE_INTEGER,
            limit: TxnFilter.DEFAULT_TXN_LIMIT,
            name: name,
            accountId: null,
            contextId: contextId,
        });
    }
}
